---
title: "`r paste0(config::get(file = 'Step2_config.yml')$PT_ID, '_CNV_analysis')`"
author: "`r config::get(file = 'Step2_config.yml')$document$author`"
date: "`r Sys.Date()`"
output:
  pdf_document:
    keep_tex: true
    latex_engine: pdflatex
    fig_caption: yes
    highlight: haddock
    number_sections: yes

knit: (
  function(inputFile, encoding) {
    config=config::get(file = 'Step2_config.yml') 
    
    pSubTitle <- paste0(config$PT_ID,"_",config$Sample_input,"_CNV.pdf")
    
    base_output_dir=file.path(config$Path,"Report")
    
    DIR=file.path(base_output_dir, config$PT_ID)
    
    rmarkdown::render( 
      input       = inputFile, 
      encoding    = encoding, 
      output_file = pSubTitle,
      output_dir =DIR )})
---

```{r,include=F}
rm(list=ls())
```


```{r,setup, include=FALSE}

# 
knitr::opts_chunk$set(echo = F,warnings =F, message=F,include=F)
options(knitr.table.format = 'markdown',encoding = 'UTF-8',warnings =F, message=F ) 
# Check and create output directory if it doesn't exist
config <- config::get(file = 'Step2_config.yml')
Path <- config$Path
base_output_dir=file.path(Path,"Report")
output_dir <- file.path(base_output_dir, config$PT_ID)
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

```

```{r}

source(file.path(getwd(),"Rfunction.R"))

```


```{r}
library(dplyr)
library(ggplot2)
```


```{r,include=F}
PT_ID <- config$PT_ID

Email <- config$Email
Reference_path <- config$Reference_path
Sample_input=config$Sample_input%>%as.integer()
Rcode_path=file.path(Path,"Rcode",PT_ID)
Rdata_path=file.path(Path,"Rdata",PT_ID )
Rawdata_path=file.path(Path,"Rawdata",PT_ID)
Process_path=file.path(Path,"Processed")
Public_database_path="/Volumes/lyu.yang/MAC_1/R/Public_database"
Reference_path="/Volumes/lyu.yang/MAC_1/R/Public_database/Reference"
```

```{r}

base_output_dir=file.path(Path,"Report")
output_dir <- file.path(base_output_dir, config$PT_ID)
pSubTitle <- paste0(PT_ID,"_",Sample_input,"_CNV.pdf")
output_file=file.path(output_dir,pSubTitle )

if (file.exists( output_file)) {
  quit()
}
```


```{r,include=F}
#download Chrosmome_cytobank.rdata from synapse
if (!file.exists(file.path(Reference_path, "Chrosmome_cytobank.rdata") )) {
  Password=read.table(PW_path)%>%as.character()
  synapser::synLogin(email = Email, password = Password)
  synapser::synGet("syn62173943",downloadLocation = Reference_path,ifcollision="keep.local")
}
load("/Volumes/lyu.yang/MAC_1/R/Public_database/Reference/Chrosmome_cytobank.rdata")

```

```{r}
# The data from step1
dir(Process_path)
load(file.path(Process_path,PT_ID,paste0(PT_ID,"_Germline_somatic_RD_BAF.rdata")))

cgwtools::lsdata(file.path(Process_path,PT_ID,paste0(PT_ID,"_Germline_somatic_RD_BAF.rdata")))

 
```

```{r}
Tumor_samples=PT_Sample_info%>%dplyr::filter(tumorType!="Normal")
RDR_BAF_samples=names(Somatic_CNV)
Sample_name=RDR_BAF_samples[Sample_input]
```

```{r,include=T}
knitr::kable(Sample_name)%>%kableExtra::kable_styling()
```

```{r}
categorize_CNV <- function(data) {
  data <- data %>%
    rowwise() %>%
    mutate(CNV = ifelse(
      (Norm_RDR >= normal_setting$Norm_RDR[1] & Norm_RDR <= normal_setting$Norm_RDR[2]) & (BAF >= normal_setting$BAF[1] & BAF <= normal_setting$BAF[2]), "Normal",
      ifelse((Norm_RDR >= LOH_setting$Norm_RDR[1] & Norm_RDR <= LOH_setting$Norm_RDR[2]) & (BAF >= LOH_setting$BAF[1] & BAF <= LOH_setting$BAF[2]), "LOH",
             #ifelse(Norm_RDR >= trisomy_setting$Norm_RDR[1] & Norm_RDR <= trisomy_setting$Norm_RDR[2] & BAF >= trisomy_setting$BAF[1] & BAF <= trisomy_setting$BAF[2], "Trisomy",
             ifelse(Norm_RDR >= trisomy_setting$Norm_RDR[1] & Norm_RDR <= trisomy_setting$Norm_RDR[2] & BAF<=0.3, "Trisomy", # Trisomy+LOH
                    ifelse(Norm_RDR >= tetrasomy_setting$Norm_RDR[1] & Norm_RDR <= tetrasomy_setting$Norm_RDR[2] & BAF >= tetrasomy_setting$BAF[1] & BAF <= tetrasomy_setting$BAF[2], "Tetrasomy",
                           ifelse(Norm_RDR >= tetrasomy_LOH_setting$Norm_RDR[1] & Norm_RDR <= tetrasomy_LOH_setting$Norm_RDR[2] & BAF >= tetrasomy_LOH_setting$BAF[1] & BAF <= tetrasomy_LOH_setting$BAF[2], "Tetrasomy_LOH",
                                  ifelse(Norm_RDR >= het_loss_setting$Norm_RDR[1] & Norm_RDR <= het_loss_setting$Norm_RDR[2] & BAF >= het_loss_setting$BAF[1] & BAF <= het_loss_setting$BAF[2], "Het_loss",
                                         ifelse(Norm_RDR >= homo_loss_setting$Norm_RDR[1] & Norm_RDR <= homo_loss_setting$Norm_RDR[2] & (BAF >= homo_loss_setting$BAF[2] | BAF <= homo_loss_setting$BAF[1]), "Homo_loss",
                                                "Unclassified"))))))))
  
  return(data)
}
```


```{r,include=F}
dir(Public_database_path)
dir.create(Rdata_path)
dir.create(Rawdata_path)
dir.create(Process_path)
#output_dir="/Volumes/lyu.yang/MAC_1/R/Project/14_MPNST_tumor_evolution/Output/JH-2-002"
# download Hg38_gene_chr_position.csv from synapse syn26376828" in your reference pathay

```

# Germline CNV 
```{r,include=F}
Germline_CNV_ALT=Germline_CNV%>%mutate(Sample=gsub("_segments.tsv","" ,Filename))%>%mutate(CN_name=ifelse(grepl("NF1",Gene_list),"NF1" ,gsub(",.*","" ,Gene_list)))%>%mutate(CN_define=ifelse(CN==1,"microdeletion",ifelse(CN==0,"Homo_deletion",ifelse(CN==3,"Trisomy",ifelse(CN=4,"Tetrasomey",CN))))) %>%   mutate(name=paste0(CN_name,"_",CN_define))%>%.[,c("Sample","Chromosome","Start","End","ALT","CN","Gene_list","name" )]
colnames(Germline_CNV_ALT)=c("sample","chr","start","end","alt","cn","genelist","name")

```


```{r,include=T}
if (nrow(Germline_CNV_ALT)>0) {
 knitr::kable(Germline_CNV_ALT,caption = "Germline structure variant",longtable = TRUE) %>%kableExtra::kable_styling(latex_options =  c("scale_down","hold_position"))%>%kableExtra:: column_spec(7, width = "3cm")%>%kableExtra:: column_spec(1, width = "1.5cm")
} else {
  print("No germline CNV was found")
  
} 

```

```{r}
Somatic_seg=Somatic_CNV[[Sample_name]]%>%Gene_annotation_for_CNV_2(.)
```

## Step1:  Define normal CN AND calcuate RDR bias 

```{r,include=T}
if (max(Somatic_seg$RDR)>3) {
  
  Data_filter=Somatic_seg%>%dplyr::filter(RDR<3)
  Outlier=Somatic_seg%>%dplyr::filter(RDR>=3)%>%mutate(Gene_list=gsub(",", ", ", Gene_list))%>%mutate(Length=End-Start)
  knitr::kable(Outlier,caption = "Seg with RDR >2.5")%>%kableExtra::kable_styling(latex_options = c("scale_down")  )%>%kableExtra:: column_spec(2,9, width = "3cm")

} else{
  
  Data_filter=Somatic_seg
}

```

```{r,include=F,eval=F}

# Kmeans=kmeans(Data_filter[,c("RDR","BAF")],centers=10,nstart=30,algorithm ="MacQueen" )
# Kmeans_anno=Kmeans[[1]]%>%as.data.frame()%>%setNames(c("Cluster") )%>%mutate(RowID=rownames(.))
# Kmeans_cluster=Kmeans[[2]]%>%as.data.frame() %>%arrange(RDR)%>%mutate(Cluster=as.integer(rownames(.)))%>%mutate(CNV=ifelse(RDR>0.8&RDR<1.1&BAF>0.45,"Normal",ifelse(RDR<0.5&(BAF>0.45|BAF<0.1),"Homo_loss","Other" )))
# # set normal RDR manually
# Normal_Chr=Kmeans_cluster%>%dplyr::filter(RDR>0.85&BAF>0.45&RDR>0.85&RDR<1)
# RDR_bias=Normal_Chr$RDR%>%round(.,2)
# Data_Kmeans_anno=Data_filter%>%mutate(RowID=rownames(.))%>%left_join(.,Kmeans_anno)%>%mutate(Type=ifelse((BAF==0|BAF>0.45)&RDR<0.7,"Homo_loss","Other"))
# Kmeans_cluster_normal=Kmeans_cluster%>%dplyr::filter(CNV=="Normal")
# knitr::kable(Kmeans_cluster_normal,caption ="Normal CNV defination")%>%kableExtra::kable_styling(latex_options = c("scale_down","hold_position") )
# ggplot(Data_Kmeans_anno, aes(x=BAF, y=RDR,color=as.character(Cluster))) + geom_point(size=3)+ylim(c(0,3)) +ggtitle("Raw_RDR_BAF")
```

# Tumor copy number balance checking
```{r,include=T,message=F}
# Tumor clonality based on clustering
#Kmenas for clustering seg data after removing germline CNV**
#Read depth data from GATK**
#Define gain and loss by Kmeans**
#Tumor_purity: "1-RDR_homo_deletion/Bias" **
#Ploidy: "1/Normal_RDR; RDR_bias=Normal_RDR"**
create_bins <- function(start, end, bin_size) {
  seq(from = start, to = end, by = bin_size)
}



Data_fileter_bin=Data_filter[,c("Chromosome","Start","End","RDR","BAF")]%>%
  mutate(bin_starts = purrr::map2(Start, End, ~create_bins(.x, .y, 50000))) %>%
  tidyr::unnest(bin_starts) %>%
  mutate(bin_ends = pmin(bin_starts + 9999, End))

Out_all=density_with_hdi (Data_fileter_bin$RDR,bandwidth = 0.03, credMass = 0.9)
Max_out_all=Out_all[[2]]%>%dplyr::filter(Y==max(Y))
if (Max_out_all$Value>0.95 & Max_out_all$Value<1.05) {
  Max_out=Max_out_all
  Bias_range=c(Max_out$min_value,Max_out$max_value)
} else {
  
  Data_fileter_bin_filter=Data_fileter_bin%>%dplyr::filter(RDR>0.7&RDR<1.35)
  #Bandwidth_RDR=
 Out=density_with_hdi (Data_fileter_bin_filter$RDR,bandwidth = 0.03, credMass = 0.9)

 # if the main peak is not located around 1, that means read depth has bias.
 Max_out=Out[[2]]%>%dplyr::filter(Y==max(Y))

 Bias_range=c(Max_out$min_value,Max_out$max_value)
}





print(paste0("peak:",Max_out$Value))
```


```{r,fig.width=16,fig.height=18,include=T}
if (config$tumor_type == "PDX"|grepl("PDX",Sample_name)) {
  Genomic_settings <- config$genomic_settings$PDX
} else {
  Genomic_settings <- config$genomic_settings$tumor
}
 



Cluster_number=config$Cluster_number
if (Max_out$Value<0.95) {
 
 print ("bias")
 UMAP_data=Data_filter[,c("RDR","BAF")] 
# Need adjust cluster number in yml based on the variety of data.
 
# Determine settings based on tumor type

 P1=generate_umap_plots(Data_filter,UMAP_data, n_cluster = Cluster_number, seed = 123456)
 
 Normal_cluster=P1$summary_table%>%dplyr::filter(BAF>0.45&RDR<0.95&RDR>Bias_range[1])%>%mutate(Diff=abs(RDR-1))%>%arrange(Diff)%>%.$Cluster%>%.[1]
 if (! is.na(Normal_cluster)) {
   P2=generate_normal_plots(Data_filter,UMAP_data, n_cluster =Cluster_number, seed = 123456,Cluster_select=Normal_cluster,Bias="")
   cowplot::plot_grid(P1$UMAP_plot_list,P2$UMAP_plot_list,ncol=1)
 }  else {
   
   print("no normal cluster was found, please change the kmeans cluster number in your yml file ")
   knitr::knit_exit()
   
 }
   
}  else if (Max_out$Value>1.05) {
 print ("bias")
 UMAP_data=Data_filter[,c("RDR","BAF")] 
# Need adjust cluster number in yml based on the variety of data.
 P1=generate_umap_plots(Data_filter,UMAP_data, n_cluster = Cluster_number, seed = 123456)
 
 Normal_cluster=P1$summary_table%>%dplyr::filter(BAF>0.45&RDR>1&RDR<Bias_range[2])%>%mutate(Diff=abs(RDR-1))%>%arrange(Diff)%>%.$Cluster%>%.[1]
 if (! is.na(Normal_cluster)) {
   P2=generate_normal_plots(Data_filter,UMAP_data, n_cluster =Cluster_number, seed = 123456,Cluster_select=Normal_cluster,Bias="")
   cowplot::plot_grid(P1$UMAP_plot_list,P2$UMAP_plot_list,ncol=1)
 }  else {
   
   print("no normal cluster was found, please change the kmeans cluster number in your yml file ")
   knitr::knit_exit()
   
 }
   
}
```




## Clonality defniation by UMAP-based kmeans 

```{r}
#setting
Type_color=RColorBrewer::brewer.pal(9,"Set1")[1:6]
names(Type_color)=factor(c("Homo_loss","Het_loss","Normal","Trisomy","Tetrasomy","Other"),levels= c("Homo_loss","Het_loss","Normal","Trisomy","Tetrasomy","Other"))
CN_label=data.frame(Type=c("Homo_loss","Het_loss","Normal" ,"Trisomy" , "Tetrasomy"),CN=c(0,1:4))
CNV_colors <-RColorBrewer::brewer.pal(9,"Paired")
names(CNV_colors)=c("Normal","LOH" , "Trisomy","Trisomy_Tetrasomy_mix" ,"High_level_ampli", "Tetrasomy", "Tetrasomy_LOH" ,"Homo_loss" ,"Het_loss" )
normal_setting=Genomic_settings$normal
LOH_setting <- Genomic_settings$LOH
trisomy_setting <- Genomic_settings$trisomy
tetrasomy_setting <- Genomic_settings$tetrasomy
tetrasomy_LOH_setting <- Genomic_settings$tetrasomy_LOH
het_loss_setting <- Genomic_settings$het_loss
homo_loss_setting <- Genomic_settings$homo_loss


```



```{r}
# trisomy_model <- lm(BAF ~ RDR, data = trisomy_data)
tumor_purity <- seq(0.2, 0.9, by = 0.1)
LOH_per <- seq(0, 1, by = 0.2)
LOH_type <- c("Gain", "Loss")
Tetrasomy_data <- data.frame(Tumor_Purity = tumor_purity, RDR =1+tumor_purity , BAF = 0.5)

# Create a grid of all combinations of tumor_purity, LOH_per, and LOH_type
Tetrasomy_LOH_data <- expand.grid(Tumor_Purity = tumor_purity, 
                                LOH_per = LOH_per, 
                                LOH_type = LOH_type)%>%mutate(LOH_purity=Tumor_Purity*LOH_per)

# Calculate RDR for Tetrasomy
Tetrasomy_LOH_data$RDR <- 1+Tetrasomy_LOH_data$Tumor_Purity

Tetrasomy_LOH_data2 <- Tetrasomy_LOH_data %>%
  mutate(BAF = case_when(
    LOH_type == "Loss" ~ 0.5 * (1 - Tumor_Purity*LOH_purity) + 0.25 *  (Tumor_Purity*LOH_purity) ,
    LOH_type == "Gain" ~ 0.5 * (1 - Tumor_Purity*LOH_purity) + 1 *  (Tumor_Purity*LOH_purity) ,
    TRUE ~ NA_real_  # Default case, should not be needed but included for safety
  ))%>%mutate(BAF=ifelse(BAF>0.5,1-BAF,BAF))%>%
  mutate(LOH_Label = paste( "LOH Type:", LOH_type, 
                       "LOH %:", LOH_per, 
                       sep = " "))


Tetrasomy_LOH_data_line_model=Tetrasomy_LOH_data2%>%dplyr::filter(LOH_per==1)%>%dplyr::filter(LOH_type=="Gain")
Tetrasomy_LOH_model <- lm(BAF ~ RDR, data = Tetrasomy_LOH_data_line_model)



if (Max_out$Value<0.95 | Max_out$Value>1.05) {
 df=P2[[2]]
 Bias_set=P2[[3]]
} else {
  df=Data_filter%>%mutate(Norm_RDR=RDR)
  Bias_set=1
  
}


Data_select=df%>%dplyr::filter(Norm_RDR>1.1&Norm_RDR<1.45&BAF>0.3&BAF<0.5)

if (grepl("PDX",Sample_name)){
  trisomy_data <- data.frame(
  Tumor_Purity = seq(0.4, 1, by = 0.1),
  RDR = 1 + seq(0.4, 1, by = 0.1) / 2,
  BAF = 0.33 * seq(0.4, 1, by = 0.1) + 0.5 * (1 - seq(0.4,1, by = 0.1))
)

 Data_select=df%>%dplyr::filter(Norm_RDR>1.1&Norm_RDR<1.5&BAF>0.3&BAF<0.5)
  
}


if (nrow(Data_select)>0) {
 results <- Data_select%>%.[,c("Chromosome","Start", "End","BAF","Norm_RDR")]%>%dplyr::rename(RDR=Norm_RDR) %>%
   rowwise() %>%
   mutate(
    Analysis = list(Trisomy_model_predict(Tetrasomy_LOH_model, RDR, BAF, threshold = 0.03))
  )%>%ungroup() %>%  # Optionally remove old columns if you only want the results
  tidyr::unnest_wider(col = Analysis)%>%arrange(RDR)

 Data_select_2=Data_select%>%left_join(.,results[,c(1,2,3,9)])%>%mutate(CNV=ifelse(Is_Close_To_Line==T|BAF>0.45,"Tetrasomy","Other"))
 G1=ggplot(Data_select_2,aes(x=BAF,y=Norm_RDR,color=factor(Is_Close_To_Line)))+geom_point()+geom_point(data=Tetrasomy_LOH_data_line_model,aes(x=BAF,y=RDR),shape=23, fill="blue", color="darkred", size=3)
  coef_intercept <- coef(Tetrasomy_LOH_model)[1]
   coef_slope <- coef(Tetrasomy_LOH_model)[2]
  # Adding line calculation to results
   # Data_select_3 <- Data_select_2 %>%
   #    mutate(Line_BAF = coef_intercept + coef_slope * Norm_RDR) %>%
   #     mutate(CNV2=ifelse(BAF < Line_BAF,"Trisomy",CNV) )%>%dplyr::select(-Line_BAF) # Select points where BAF is less than the calculated
   Data_select_3 <- Data_select_2%>% dplyr::filter(CNV=="Other")%>%
       mutate(Line_BAF = coef_intercept + coef_slope * Norm_RDR) %>%
        mutate(CNV=ifelse(BAF < Line_BAF,"Trisomy","Tetrasomy") )%>%dplyr::select(-Line_BAF)
   Data_select_2_rm=Data_select_2%>% dplyr::filter(CNV!="Other")
    Data_select_4=rbind(Data_select_2_rm,Data_select_3)%>%dplyr::rename(CNV2=CNV)
   
   G2=ggplot(Data_select_4 ,aes(x=BAF,y=Norm_RDR,color=factor(CNV2)))+geom_point()+geom_point(data=Tetrasomy_LOH_data_line_model,aes(x=BAF,y=RDR),shape=23, fill="blue", color="darkred", size=3)
   
   categorized_data<- categorize_CNV(df)%>%left_join(.,Data_select_4[,c("Chromosome","Start", "End","CNV2")])%>%mutate(CNV3 =ifelse(!is.na(CNV2), CNV2, as.character(CNV)))%>%mutate(CNV3=ifelse(is.na(CNV3),CNV,CNV3))%>%dplyr::select(-CNV)%>%dplyr::rename(CNV=CNV3)

} else {
  categorized_data<- categorize_CNV(df)
}
```


```{r,include=T,fig.width=10,fig.height=6}
if (nrow(Data_select)>0) {
cowplot::plot_grid(G1,G2)
}
```



```{r,fig.width=6,fig.height=6,include=T}

out=categorized_data%>%dplyr::filter(CNV=="Unclassified")

P5 =ggplot(categorized_data, aes(x=BAF, y=Norm_RDR,color=CNV)) + geom_point(size=3)+scale_color_manual(values=CNV_colors) +ggtitle("Cluster_CNV2")
P5

```


```{r}
All_Sub_p=calculate_subclone_purities(categorized_data)
```
\newpage

## Clonality from LOH and Trisomy and Hetloss
```{r,include=T,fig.height=3,fig.width=6}
No_normal_data <-All_Sub_p%>%dplyr::filter(CNV!="Normal")

if (nrow(No_normal_data ) <15) {
  Select_data =No_normal_data
  data=Select_data$Subclone_p%>%sort()
# Example usage
 set.seed(12345)
 Bandwidth=Genomic_settings$bandwidth_1
 print(Bandwidth)
 result <- density_with_hdi(data, bandwidth = Bandwidth, credMass = 0.9)
result[[2]]
Clonaity_define=result[[2]]%>%arrange(desc(Value))%>%mutate(Group=as.character(c(1:nrow(.))))%>%dplyr::filter(Value<0.9)
  
}
```


```{r,include=T,fig.height=3,fig.width=12}
if (nrow(No_normal_data) >15) {

 Select_data <-All_Sub_p %>%dplyr::filter(CNV %in%c("LOH","Hetloss","Trisomy"))
 data=Select_data$Subclone_p%>%sort()
# Example usage
set.seed(12345)
Bandwidth=Genomic_settings$bandwidth_1

result <- density_with_hdi(data, bandwidth = Bandwidth, credMass = 0.9)
#result[[2]]
Data=All_Sub_p %>%dplyr::filter(!CNV %in%c("LOH","Hetloss","Trisomy")) %>%dplyr::filter(!is.na(Subclone_p))%>%dplyr::filter(Subclone_p<0.9)
out=group_numbers_with_clonality(Data$Subclone_p,result[[2]], threshold=0.1) 
grouped_other_df= Data%>%as.data.frame()%>%mutate(Group=out)
Other=grouped_other_df%>%dplyr::filter(Group=="other")
if(nrow(Other)>0) {
  
  data=Other$Subclone_p%>%na.omit()
  Group_number=grouped_other_df$Group%>%sort()%>%as.integer()%>%na.omit()%>%max()+1

  Bandwidth=Genomic_settings$bandwidth_2
  print(Bandwidth)
  result_other <- density_with_hdi(data, bandwidth = Bandwidth, credMass = 0.9,initial_group_number= Group_number)
 
 #result_other[[2]]
  if (!grepl("PDX",Sample_name)){
     Clonaity_define=rbind(result[[2]],result_other[[2]])%>%arrange(desc(Value))%>%mutate(Group=as.character(c(1:nrow(.))))%>%dplyr::filter(Value<0.9)
  } else {
    Clonaity_define=rbind(result[[2]],result_other[[2]])%>%arrange(desc(Value))%>%mutate(Group=as.character(c(1:nrow(.))))
  
  } 
}
 else {
   Clonaity_define=result[[2]]%>%arrange(desc(Value))%>%mutate(Group=as.character(c(1:nrow(.))))
   
 }

}

 
```


```{r}
# need to update
High_ampli_cal_CN = function(RDR, TP_list, CN_range) {
  output = data.frame()

  for (i in TP_list) {
    Normal_CN = (1 - i) * 2
    row = data.frame(TP = i)

    for (CN in CN_range) {
      CP = i * CN
      RDR_value = (Normal_CN + CP) / 2
      row[paste0("CN", CN, "_RDR")] = RDR_value
    }

    output = rbind(output, row)
  }
  output2=output%>%as.data.frame()%>%mutate(RDR_value=RDR)

  # Initialize variables to store the closest values
  closest_value = Inf
  closest_TP = NA
  closest_CN = NA
  diff_list=list()
  # Iterate through the output to find the closest value to RDR
  for (row in 1:nrow(output)) {
    for (CN in CN_range) {
      diff = abs(output[row, paste0("CN", CN, "_RDR")] - RDR)
      if (diff < closest_value) {
        closest_value = diff
        closest_TP = output$TP[row]
        closest_CN = CN
      }


    }
  }

  # Create the final output table with the closest TP and CN
  final_output = data.frame(Value = closest_TP, CN = closest_CN, RDR_diff = closest_value)%>%mutate(CN=ifelse(RDR_diff<0.25,CN,NA))

  return(final_output)
}


```

```{r}
# High level application defination
Data=All_Sub_p$Subclone_p
out <- group_numbers_with_clonality(Data,Clonaity_define, threshold=0.15)
# High level ampli defination
Data_update_9=All_Sub_p%>%as.data.frame()%>%mutate(Group=out)%>%left_join(.,Clonaity_define)%>%mutate(Group2=paste0("Clonal_",Group))%>%mutate(Group2=ifelse(CNV=="Normal","Normal",Group2))%>%mutate(CNV=ifelse(CNV=="Unclassified"&Norm_RDR>1.9,"High_level_ampli",CNV))


High_ampli_data=Data_update_9%>%dplyr::filter(CNV=="High_level_ampli")

# for the High level amplication with 6, 8, 10
High_ampli_data_baf_0.5=High_ampli_data%>%dplyr::filter(BAF>0.4)
if(nrow(High_ampli_data_baf_0.5) >0){
 Test=sapply(High_ampli_data_baf_0.5$Norm_RDR,function(x) out=unlist(High_ampli_cal_CN(x,Clonaity_define$Value,CN_range=c(6,8,10))))%>%t()%>%as.data.frame()%>%left_join(.,Clonaity_define)

 High_ampli_data2=High_ampli_data_baf_0.5%>%mutate(CN=Test$CN,Group2=Test$Group,Value=Test$Value,Group=Test$Group)

 High_ampli_data_baf_other=High_ampli_data%>%dplyr::filter(BAF<=0.45)%>%mutate(CN=">4",Group2="High_level_ampli")

 High_ampli_combine=rbind(High_ampli_data_baf_other,High_ampli_data2)
} else {
   High_ampli_combine=High_ampli_data%>%mutate(CN=">4",Group2="High_level_ampli")
  
}
#%>%left_join(.,Clonaity_define)

No_high=Data_update_9%>%dplyr::filter(CNV!="High_level_ampli")%>%mutate(Value=ifelse(is.na(Value),Subclone_p,Value)) %>%
    mutate(CN = case_when(
      CNV == "LOH" ~ 2,
      grepl("Tetrasomy",CNV) ~ 4,
       grepl("Trisomy",CNV)  ~ 3,
      CNV == "Het_loss" ~ 1,
      CNV == "Homo_loss" ~ 0,
      CNV == "Normal" ~ 2,
      TRUE ~ NA_real_  # In case CNV doesn't match any condition
    ))

# correct NF1 information 
Data_update_10=rbind(High_ampli_combine,No_high)%>%mutate(Group2=paste0(Group,":",round(Value,2)))%>%mutate(Group2=ifelse(CNV=="Normal","Normal",ifelse(Group2=="other:1","other:NA", Group2))) %>%arrange(Chromosome,Start)

Data_update_10%>%dplyr::filter(Group2=="other:NA")
```


```{r}
# if (grepl("NF1",Germline_CNV_ALT$genelist)){
#   print("true")
#   Data_update_11=Data_update_10%>%mutate(CNV=ifelse(grepl(",NF1,",Gene_list),"Germ_loss",CNV))%>%mutate(CN=ifelse(grepl(",NF1,",Gene_list),1,CN))%>%mutate(Group2=ifelse(grepl(",NF1,",Gene_list),"Germline_Variant",Group2))
#   
# }
```


```{r,include=F}
P11 =ggplot(Data_update_10, aes(x=BAF, y=Norm_RDR,color=CNV)) + geom_point(size=3)+scale_color_manual(values=CNV_colors) +ggtitle("Clonal")+facet_wrap(~Group2)
```


```{r,fig.width=12,fig.height=18,include=T}
P11
```

\newpage

```{r,fig.width=24,fig.height=6,include=T}
Somatic_seg_BAF_plot_2(Data_update_10)

```
```{r,fig.width=24,fig.height=6,include=T}
Somatic_seg_Raw_RDR_plot_2 (Data_update_10)
```


```{r,fig.width=24,fig.height=6,include=T}
Somatic_seg_Norm_RDR_plot_2(Data_update_10)
```

```{r,fig.width=24,fig.height=6,include=T}
# Example usage (assuming Somatic_SV and Germline_SV data frames are properly defined):

Germline_CNV_ALT_unique=Germline_CNV_ALT%>%distinct(name,.keep_all = T)

Data_update_10_refine=Data_update_10%>%dplyr::filter(!grepl(">4",as.character(CN)))%>%mutate(CN=as.numeric(CN))%>%dplyr::filter(CN<6.1)%>%mutate(CN=ifelse(Value<0.15&!is.na(Value),2,CN))%>%
  mutate(Group2=ifelse(Value<0.15,"Normal",Group2))%>%
  mutate(Value=ifelse(Value<0.15&!is.na(Value),1,Value))

MAX_cn=max(Data_update_10_refine$CN)
 

Somatic_CN_plot_2(Data_update_10_refine,Germline_SV =Germline_CNV_ALT_unique,MAX_CN =MAX_cn,Title=paste0("Somatic CNV plot"," Clone_proportion>0.15" ))
```

\newpage

## High level amplication 
```{r,include=T}
High_ami=Data_update_10%>%dplyr::filter(CNV=="High_level_ampli")%>%.[c("Chromosome","Gene_list","Start","End" ,"Norm_RDR","BAF","Group2","CN")]%>%mutate(Gene_list=gsub(",", ", ", Gene_list))%>%arrange(Chromosome,Start)
knitr::kable(High_ami,caption = "High amplication",longtable = TRUE) %>%kableExtra::kable_styling(latex_options =  c("scale_down","hold_position"))%>%kableExtra:: column_spec(2, width = "3cm")
```

\newpage
```{r,include=T}
NF1=Data_update_10%>%dplyr::filter(grepl(",NF1,|,NF1$",Gene_list))%>%.[c("Chromosome","Gene_list","Start","End" ,"Norm_RDR","BAF","Group2","CN")]%>%mutate(Gene_list=gsub(",", ", ", Gene_list))

if (NF1$CN!=2) {
    knitr::kable(NF1,caption = "NF1 somatic",booktabs = TRUE) %>%
      kableExtra::kable_styling(latex_options =  c("scale_down","hold_position"))%>%kableExtra::column_spec(column=2, width = "5cm")
  
  }

```
```{r,include=T}
CDKN2A=Data_update_10%>%dplyr::filter(grepl(",CDKN2A,|CDKN2A$",Gene_list))%>%.[c("Chromosome","Gene_list","Start","End" ,"Norm_RDR","BAF","Group2","CN")]%>%mutate(Gene_list=gsub(",", ", ", Gene_list))

if (CDKN2A$CN!=2) {
knitr::kable(CDKN2A,caption = "CDKN2A somatic",booktabs = TRUE) %>%kableExtra::kable_styling(latex_options =  c("scale_down","hold_position"))%>%kableExtra::column_spec(column=2, width = "5cm")
}



```

## Het and homoloss
 
```{r,include=F}
Loss=Data_update_10%>%dplyr::filter(grepl("loss", CNV) )%>%.[c("Chromosome","Gene_list","Start","End" ,"Norm_RDR","BAF","Group2","CN")]%>%mutate(Gene_list=gsub(",", ", ", Gene_list))%>%arrange(Chromosome,Start)
knitr::kable(Loss,caption = "Loss",booktabs = TRUE) %>%kableExtra::kable_styling(latex_options =  c("scale_down","hold_position"))%>%kableExtra::column_spec(column=2, width = "3cm")

```

\newpage 

# Config setting for this sample
```{r,include=T}
if (config$tumor_type == "PDX"|grepl("PDX",Sample_name)) {
 print(config$genomic_settings$PDX) 
} else {
  print(config$genomic_settings$tumor ) 
  
  }
```


```{r}
dir.create(file.path(Process_path,PT_ID))
save(Data_update_10,Data_update_10_refine,Bias_set,file=file.path(Process_path,PT_ID ,paste0(PT_ID,"_",Sample_name,"_CNV_clonality.rdata")))
```
```{r,eval=F}
output_file <- paste0(config$PT_ID, "_", config$Sample_input, "_CNV.pdf")
output_dir <- file.path(base_output_dir, config$PT_ID)

# Render the RMarkdown file with custom output settings
rmarkdown::render(
  input = "1_step2_CNV.Rmd",
  output_file = output_file,
  output_dir = output_dir)
```


```{r}
knitr::knit_exit()
```


